crates/stackable-operator/src/commons/networking.rs

```diff
@@ -1,44 +1,136 @@
-use std::{fmt::Display, ops::Deref};
+use std::{fmt::Display, net::IpAddr, ops::Deref, str::FromStr};
 
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
+use snafu::Snafu;
 
 use crate::validation;
 
-/// A validated hostname type, for use in CRDs.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+/// A validated domain name type conforming to RFC 1123, so e.g. not an IP addresses
+#[derive(
+    Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd, JsonSchema,
+)]
 #[serde(try_from = "String", into = "String")]
-pub struct Hostname(#[validate(regex(path = "validation::RFC_1123_SUBDOMAIN_REGEX"))] String);
+pub struct DomainName(#[validate(regex(path = "validation::RFC_1123_SUBDOMAIN_REGEX"))] String);
+
+impl FromStr for DomainName {
+    type Err = validation::Errors;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        validation::is_rfc_1123_subdomain(value)?;
+        Ok(DomainName(value.to_owned()))
+    }
+}
 
-impl TryFrom<String> for Hostname {
+impl TryFrom<String> for DomainName {
     type Error = validation::Errors;
 
     fn try_from(value: String) -> Result<Self, Self::Error> {
-        validation::is_rfc_1123_subdomain(&value)?;
-        Ok(Hostname(value))
+        value.parse()
     }
 }
 
-impl From<Hostname> for String {
-    fn from(value: Hostname) -> Self {
+impl From<DomainName> for String {
+    fn from(value: DomainName) -> Self {
         value.0
     }
 }
 
-impl Display for Hostname {
+impl Display for DomainName {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.write_str(&self.0)
     }
 }
 
-impl Deref for Hostname {
+impl Deref for DomainName {
     type Target = str;
 
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
+#[derive(Debug, Snafu)]
+pub enum HostNameParseError {
+    #[snafu(display(
+        "the given hostname {hostname:?} is not a valid hostname, which needs to be either a domain name or IP address"
+    ))]
+    InvalidHostname { hostname: String },
+}
+
+/// A validated hostname, which is either a [`DomainName`] or [`IpAddr`].
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+#[serde(try_from = "String", into = "String")]
+pub enum HostName {
+    IpAddress(IpAddr),
+    DomainName(DomainName),
+}
+
+impl JsonSchema for HostName {
+    fn schema_name() -> String {
+        "HostName".to_owned()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        String::json_schema(gen)
+    }
```

Comment:
This is effectively dropping the validation regex from the composite type.

Reply:
You're right. So far we don't have any regex on HostName, but instead let IpAddr::parse do it's thing and try to parse it as domain name in case it#s not an IP.
We should be able to come up with an regex that accepts a DomainName, IPv4 or IPv6 address. But do you think it's worth the effort?

Reply:
That's fair.